src/lib.rs

```diff
@@ -1,3 +1,75 @@
+//! Landlock is a security feature available since Linux 5.13.
+//! The goal is to enable to restrict ambient rights
+//! (e.g., global filesystem access)
+//! for a set of processes by creating safe security sandboxes as new security layers
+//! in addition to the existing system-wide access-controls.
+//! This kind of sandbox is expected to help mitigate the security impact of bugs,
+//! unexpected or malicious behaviors in applications.
+//! Landlock empowers any process, including unprivileged ones, to securely restrict themselves.
+//! More information about Landlock can be found in the [official website](https://landlock.io).
+//!
+//! This crate provides a safe abstraction for the Landlock system calls, along with some helpers.
+//!
+//! # Use cases
+//!
+//! This crate is especially useful for build-in application sandboxing:
+//! * Parsers hardening (e.g., archive tools, file format conversion, renderers).
+//! * (Part of) applications with limited file renaming or linking needs
+//!   (e.g., some system or network services).
+//! * Monolith applications dealing with different levels of confidentiality
+//!   (e.g., web browser, email server).
+//!
+//! # Examples
+//!
+//! Examples can be found with the [`Ruleset` documentation](Ruleset)
+//! and the [sandboxer example](https://github.com/landlock-lsm/rust-landlock/blob/master/examples/sandboxer.rs).
+//!
+//! # Current limitations
+//!
+//! This crate handles the Landlock features available since Linux 5.13
+//! and then inherits some [kernel limitations](https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#current-limitations)
+//! that will be addressed with future kernel releases:
+//! * File reparenting: renaming or linking a file to a different parent directory is always denied.
+//! * Filesystem topology modification: arbitrary mounts are always denied.
+//!
+//! # Compatibility
+//!
+//! This crate is designed to enable security as much as possible on supported systems.
```

Comment:
I found lines 37-41 difficult to understand; let me suggest the following phrasing (assuming that I understood it correctly):

> This crate is designed to enable the strictest Landlock configuration for the the given kernel on which the program runs.
> The crate will determine compatibility based on the currently running kernel's features,
> not based on the specific rule provided by the caller.
> This way, callers can distinguish between Landlock compatibility issues inherent to the system
> and misconfigured rules (e.g., providing file names that don't exist).

I tried formulating this in active instead of passive voice above, but didn't know the code well enough to be specific; maybe "the crate" would better be "the function Xyz()" or "the class Foo" in some places?

I'm not sure I understood this in its entirety, but it seems that the core point is that callers can distinguish between issues of their own fault (wrong file names?), and issues inherent to the system they run on (missing Landlock support in kernel). Maybe this would be less abstract and easier to explain by having a section that is listing possible error scenarios, for example:

> # Errors
>
> The following errors are possible:
>
> * The kernel does not support the requested Landlock features. In best effort mode, this crate will use the best available security settings (possibly none). Otherwise, the function Foobar() returns error code XYZ
> * The file names used when constructing the rule cannot be opened. This is exposed as PathFdError when bulding a PathFd object using PathFd::new(). (?)

There's a good chance I'm overlooking some parts here :)

Reply:
Sounds good to me.
(MIcro nit, I'd say "In the default [best effort](Compatible::set_best_effort) mode, ..." to clarify that it's the default)